src/RobustLeastSquares.jl

```diff
@@ -1,5 +1,307 @@
+
+
+# Weight functions
+# ----------------
+# A good overview of these can be found at:
+# http://research.microsoft.com/en-us/um/people/zhang/INRIA/Publis/Tutorial-Estim/node24.html
+
+# Standard L2 estimator
+"""
+An m-estimator is a cost/loss function used in modified (weighted) least squares
+problems of the form:
+    min ∑ᵢ ρ(rᵢ)
+"""
+abstract MEstimator
+abstract MEstimatorConvex <: MEstimator
+abstract MEstimatorNonConvex <: MEstimator
+
+# Default implementations from definintion of cost-function rho
+"The cost (a.k.a. loss) function ρ for the M-estimator"
+function m_estimator_rho end
+"The derivative of the cost (a.k.a. loss) function ψ for the M-estimator"
+function m_estimator_psi end
+# TODO m_estimator_psi(r,est::MEstimator) = first derivative of m_estimator_rho(r,est) w.r.t. r
+"The weight function, w, for the M-estimator, to be used for modifying least-square problems"
+m_estimator_weight(r,est::MEstimator) = m_estimator_psi(r,est) ./ r
+"The square root of the weight function, sqrt(w), for the M-estimator, to be used for modifying the normal equations of a least-squares problem"
+m_estimator_sqrtweight(r,est::MEstimator) = sqrt(m_estimator_weight(r,est))
+
+"The (convex) L2 M-estimator is that of the standard least squares problem."
+immutable L2Estimator <: MEstimatorConvex; end
+L2Estimator(width) = L2Estimator()
+m_estimator_rho(r,::L2Estimator) = 0.5*r.^2
+m_estimator_psi(r,::L2Estimator) = r
+m_estimator_weight(r,::L2Estimator) = ones(size(r))
+m_estimator_sqrtweight(r,::L2Estimator) = ones(size(r))
+
+# Standard L1 estimator
+"The standard L1 M-estimator takes the absolute value of the residual, and is convex but non-smooth."
+immutable L1Estimator <: MEstimatorConvex; end
+L1Estimator(width) = L1Estimator()
+m_estimator_rho(r,::L1Estimator) = abs(r)
+m_estimator_psi(r,::L1Estimator) = sign(r)
+m_estimator_weight(r,::L1Estimator) = 1.0 ./ abs(r)
+m_estimator_sqrtweight(r,::L1Estimator) = abs(r) .^ (-0.5)
+
+# Estimator interpolating between L1 and L2 norms (analytically smooth) TODO Add width function
+"The convex L1-L2 estimator interpolates smoothly between L2 behaviour for small residuals and L1 for outliers."
+immutable L1L2Estimator <: MEstimatorConvex; width::Float64; end
+m_estimator_rho(r,::L1L2Estimator) = 2.0*(sqrt(1.0 + 0.5*r.*r)-1.0)
+m_estimator_psi(r,::L1L2Estimator) = r ./ sqrt(1 + 0.5*r.*r)
+m_estimator_weight(r,::L1L2Estimator) = 1.0 / sqrt(1+0.5*r.*r)
+m_estimator_sqrtweight(r,::L1L2Estimator) = (1+0.5*r.*r) .^ (-1/4)
+
+# Huber Estimator for M-estimation
+# C2-continuous, behaving as L1 at long distances and L2 at short
+"The convex Huber estimator switches from between quadratic and linear cost/loss function at a certain cutoff."
+immutable HuberEstimator <: MEstimatorConvex; width::Float64; end
+function m_estimator_rho(r,est::HuberEstimator)
+    rho = 0.5*r.^2
+    absr = abs(r)
+    rho[absr .> est.width] = est.width * (absr[absr .> est.width] - 0.5*est.width)
+    return rho
```

Comment:
Does it ever come up that you ever actually want \rho and \psi for an application? I'm guessing maybe if you are using an optimisation routine that doesn't use weights (Nelder-Mead simplex type thing)?